src/bridge/crypto/rsa.cc

```diff
@@ -107,28 +134,38 @@ int Sign(int type, const unsigned char *digest_bytes,
   // conversion functions refer to some OpenSSL API functions.
   KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
       const CryptoKeyHandle *crypto_key_handle,
-      GetCryptoKeyHandleFromOpenSslRsa(rsa), false);
+      GetCryptoKeyHandleFromOpenSslRsa(rsa), 0);
   KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
       const DigestCase digest_type,
-      ConvertOpenSslNidToDigestType(type), false);
+      ConvertOpenSslNidToDigestType(type), 0);
   KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
       const std::string digest_string,
-      MakeDigestString(digest_bytes, digest_length), false);
+      MakeDigestString(digest_bytes, digest_length), 0);
 
   // Delegate handling of the signing operation to the backing layer.
   KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
       const std::string signature,
-      crypto_key_handle->Sign(digest_type, digest_string), false);
+      crypto_key_handle->Sign(digest_type, digest_string), 0);
 
   // Copy results into the return pointers.
+  if (signature_length != nullptr) {
+    *signature_length = signature.length();
+  }
   if (signature_return != nullptr) {
```

Comment:
It's a common pattern in OpenSSL for `signature_return` to be `nullptr`---what the docs suggest is doing something like this to first get the `signature_length` to know how much memory to allocate:

https://github.com/googleinterns/cloud-kms-oss-tools/blob/a69e60ca3788624db9b4083e5713723ee4e2d868/src/bridge/key_loader/key_loader_test.cc#L293-L309

Some code that I've seen online also does the first `RSA_sign` call as `signature_return == nullptr` and the second `RSA_sign` call as `signature_length == nullptr`, so both of these situations can happen.

Reply:
Can you given an external example? That generally makes no sense because it means that an expensive signature operation would be done twice.

Reply:
See https://wiki.openssl.org/index.php/EVP_Signing_and_Verifying#Signing. 

From my own tests, it looks like that the OpenSSL wrappers do some caching and so the signing operation doesn't actually get called twice.

Reply:
Interestingly, in the reference you provided, it says:
> One gotcha to be aware of is that the first call to EVP_DigestSignFinal simply returns the maximum size of the buffer required. This will not always be the same as the size of the generated signature (specifically in the case of DSA and ECDSA). This means that you should also take account of the value of the length returned on the second call (in the slen variable in this example) when making use of the signature.

This seems to imply that:
1. if `buf == nullptr` and `len != nullptr`, it's ok to just set `*len = RSA_size` and just not do the signature.
2. If `buf != nullptr` and `len == nullptr`, we could still return an error. It seems unsafe to do otherwise, unless there's an example of this particular case.
3.`buf == nullptr` and `len == nullptr`, we can also return an error because this is clearly wrong.

I'd do the following:
- Check `len == nullptr` and error out if true
- Check `RSA_size <= 0` and error out if true
- Check `buf == nullptr` and if so `*len = RSA_size` and return
- Do signature
- Check that `sig.length() > RSA_size` and error out if true
- Copy signature bytes
- Set `*len = sig.length()`